### PR TITLE
feat: Add initial implementation of data analysis tools

### PR DESCRIPTION
**Pull request by AnkitRajSri**
I have added the muser-data-app project to the branch of the same name.

The app currently offers the below functionalities:
• The app is now able to extract data from Spotify API, save a raw CSV file with the timestamp, and perform ETL operation to dump the extracted data in SQL Server table.
• In addition to the data extraction functionality, we have incorporated functionality to build a doc2vec NLP model and train it on the data dumped in the SQL table.
• There is one final functionality to expand the muser data, collected from FireBase, with metadata information collected from Spotify. The app first queries Spotify for an exact match with the muser record (artist, track, and album), and in case there is no exact match, the app utilizes the doc2vec model to predict the most similar match in the SQL table.

We are still working to make the UI more interactive.

### CODE REVIEW

File: AI/spotifydataextractor.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
+import time
+import numpy as np
+import pandas as pd
+import os
+
```
Comment (barbeau) [inline AI/spotifydataextractor.py:23]: Needs comments/documentation for what this script does

File: AI/muserdatabuilder.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import pandas as pd
+import os
+from AI.models import NLPModel
+
+os.chdir(r'C:\Users\sriva\Desktop\edu.usf.sas.pal.muser\SpotifyDataExtractor')
+
```
Comment (barbeau) [inline AI/muserdatabuilder.py:25]: Needs comments/documentation for what this script does

File: AI/models.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
+import pandas as pd
+from textslack.textslack import TextSlack
+from gensim.models import doc2vec
+import os
+
```
Comment (barbeau) [inline AI/models.py:23]: Needs comments/documentation for what this script does

File: AI/models.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+os.chdir(r'C:\Users\sriva\Desktop\edu.usf.sas.pal.muser\SpotifyDataExtractor')
```
Comment (barbeau) [inline AI/models.py]: This path is hard-coded and won't work on someone else's computer. Could you please change to a relative project path that will work across different computers?

File: AI/__init__.py [hunk 5]
```diff
@@ -0,0 +1 @@
+
```
Comment (barbeau) [inline AI/__init__.py]: Does this file do anything? If so it needs comments

File: AI/ETL.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import os
+os.chdir(r'C:\Users\sriva\Desktop\edu.usf.sas.pal.muser\SpotifyDataExtractor')
+
```
Comment (barbeau) [inline AI/ETL.py]: Needs comments/documentation for what this script does

File: AI/ETL.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import os
+os.chdir(r'C:\Users\sriva\Desktop\edu.usf.sas.pal.muser\SpotifyDataExtractor')
```
Comment (barbeau) [inline AI/ETL.py]: Another hard-coded path

File: AI/muserdatabuilder.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
+import pandas as pd
+import os
+from AI.models import NLPModel
+
+os.chdir(r'C:\Users\sriva\Desktop\edu.usf.sas.pal.muser\SpotifyDataExtractor')
```
Comment (barbeau) [inline AI/muserdatabuilder.py]: Another hard-coded path

File: AI/muserdatabuilder.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
+import pandas as pd
```
Comment (barbeau) [inline AI/muserdatabuilder.py:20]: Every file should have an Apache v2.0 license header with "Copyright (C) 2020 University of South Florida"

File: app/__init__.py [hunk 10]
```diff
@@ -0,0 +1,13 @@
+from flask import Flask
+
```
Comment (barbeau) [inline app/__init__.py:5]: Needs comments/description of what script does

File: app/connectionmanager.py [hunk 11]
```diff
@@ -0,0 +1,25 @@
+import os
+from dotenv import load_dotenv
+import spotipy
+from spotipy.oauth2 import SpotifyClientCredentials
+from sqlalchemy import create_engine
+
+project_folder = os.path.expanduser(r'C:\Users\sriva\Desktop\edu.usf.sas.pal.muser\SpotifyDataExtractor')
```
Comment (barbeau) [inline app/connectionmanager.py]: More hard-coded paths

File: app/connectionmanager.py [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+load_dotenv(os.path.join(project_folder, '.env'))
+os.chdir(r'C:\Users\sriva\Desktop\edu.usf.sas.pal.muser\SpotifyDataExtractor')
```
Comment (barbeau) [inline app/connectionmanager.py]: More hard-coded paths

File: app/connectionmanager.py [hunk 13]
```diff
@@ -0,0 +1,25 @@
+import os
+from dotenv import load_dotenv
+import spotipy
+from spotipy.oauth2 import SpotifyClientCredentials
+from sqlalchemy import create_engine
```
Comment (barbeau) [inline app/connectionmanager.py:23]: Needs comments/description of script

File: app/view.py [hunk 14]
```diff
@@ -0,0 +1,65 @@
+from app import app
+from flask import Flask, request, render_template
+#app = Flask(__name__)
+
```
Comment (barbeau) [inline app/view.py:18]: Need description of what script does

File: app/view.py [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+import os
+from app.connectionmanager import ConnectionManager
+from AI.spotifydataextractor import SpotifyDataExtractor
+from AI.ETL import ETL
+from AI.models import NLPModel
+from AI.muserdatabuilder import MuserDataBuilder
+
+os.chdir(r'C:\Users\sriva\Desktop\edu.usf.sas.pal.muser\SpotifyDataExtractor')
```
Comment (barbeau) [inline app/view.py]: More hard-coded paths

File: config.py [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+    Use this class to share any default attributes with any subsequent
+    classes that inherit from Config.
+    """
+    DEBUG = False
+    TESTING = False
+    
+    # Only required when using the session object
+    # Generated with secrets.token_urlsafe(16)
+    # You could also use os.urandom(16)
+    SECRET_KEY = "muser_data_app"
```
Comment (barbeau) [inline config.py]: What is this key for? Any credentials used to access hosted systems shouldn't be committed in version control.

File: env [hunk 17]
```diff
@@ -0,0 +1,4 @@
+CLIENT_ID = "b7d43ed648454f8586d466bec5fc250b"
```
Comment (barbeau) [inline env]: This looks like a file that shouldn't be checked into version control for security reasons (you'll also need to change these credentials now that they've been made public). Please look at a way to define these locally and read into the project without committing them, and add instructions for how to set this up in the README.

**Review by barbeau — CHANGES_REQUESTED** [review]
Thanks @AnkitRajSri! Some comments are in-line below.

@pradeepsalunke after @AnkitRajSri makes changes to the hard-coded paths, could you please try cloning this project and running it from the command line? See the bottom of the PR for instructions how to do this. I want to make sure that anyone can clone this project and run it on their own machine without needing to edit any files (e.g., fixing hard-coded paths).

Also, the project needs a README in the root of the project that explains what the project does and how to run it. Please see https://github.com/CUTR-at-USF/muser-firebase-export for an example.

All the `.pyc` files should also be ignored and not included in the Git repository because they are generated by the Python compiler. You can add them to the `.gitignore`, either in the root directory or another one in the sub-directory. 

You can make these changes to your local branch on your computer, commit them, and then push to GitHub and they will show up in this PR.

**Comment by pradeepsalunke** [comment]
Hi Sean,
We have closed that pull request as it was not cloned from main  and raised a new and corrected one.
Regards,Pradeep
Pradeep SalunkeGraduate Student,Masters in Business Analytics and Information SystemUniversity of South Florida

**Comment by barbeau** [comment]
@AnkitRajSri Either way works - I'd prefer to continue in this pull request since I already left review comments here. So please re-open this and you'll just need to push new commits to the branch on your fork repository at `AnkitRajSri` to update it - https://github.com/AnkitRajSri/muser-data-analysis/tree/muser-data-app.

**Comment by AnkitRajSri** [comment]
@barbeau Is it okay to continue working from the forked repository on my profile, as I just created it for testing and the pull request was created from that repository by mistake?

**Comment by barbeau** [comment]
>s it okay to continue working from the forked repository on my profile, as I just created it for testing and the pull request was created from that repository by mistake?

Yes, that's fine, as long as all the changes are merged back into this project via pull requests.

**Comment by AnkitRajSri** [comment]
Reopening the pull request as per Sean's suggestion.

File: AI/muserdatabuilder.py (hunk 2 above)
Comment (AnkitRajSri) [reply AI/muserdatabuilder.py:25]: Comment implemented

File: AI/spotifydataextractor.py (hunk 1 above)
Comment (AnkitRajSri) [reply AI/spotifydataextractor.py:23]: Comment implemented

File: AI/models.py (hunk 3 above)
Comment (AnkitRajSri) [reply AI/models.py:23]: Comment implemented

File: AI/__init__.py (hunk 5 above)
Comment (AnkitRajSri) [reply AI/__init__.py]: Comment implemented

File: AI/ETL.py (hunk 7 above)
Comment (AnkitRajSri) [reply AI/ETL.py]: Comment implemented

File: AI/muserdatabuilder.py (hunk 9 above)
Comment (AnkitRajSri) [reply AI/muserdatabuilder.py:20]: Comment implemented

File: app/connectionmanager.py (hunk 11 above)
Comment (AnkitRajSri) [reply app/connectionmanager.py]: Comment implemented

File: app/connectionmanager.py (hunk 13 above)
Comment (AnkitRajSri) [reply app/connectionmanager.py:23]: Comment implemented

File: app/view.py (hunk 14 above)
Comment (AnkitRajSri) [reply app/view.py:18]: Comment implemented

File: app/view.py (hunk 15 above)
Comment (AnkitRajSri) [reply app/view.py]: Comment implemented

File: config.py (hunk 16 above)
Comment (AnkitRajSri) [reply config.py]: Comment implemented

File: env (hunk 17 above)
Comment (AnkitRajSri) [reply env]: Comment implemented.
Added .env file in gitignore.

**Comment by AnkitRajSri** [comment]
Hey Sean,

I have implemented all the comments, could you review and let me know if it needs any further updates?

Thanks,
Ankit Raj

File: config.py (hunk 16 above)
Comment (barbeau) [reply config.py]: Did something change? - I still see the credentials above. They should be removed from the committed code and if necessary read it programmatically from an unversioned file. Need information in the README for how to configure this too.

**Review by barbeau — CHANGES_REQUESTED** [review]
Thanks @AnkitRajSri. Another comment in-line.

Also, the .pyc files are still under version control - you'll need to delete them and commit that.

@pradeepsalunke Were you able to clone this repository and run it locally on your machine without changing any code?

**Comment by pradeepsalunke** [comment]
Hello team,
I was having health issues during this week and was resting and Im on the verge of recovery. I will do end to end testing and report any bugs before this weekend that is 20th December.Sorry for the inconvenience caused.Thanks Pradeep

Sent from Yahoo Mail for iPhone


On Thursday, December 17, 2020, 1:44 AM, Sean Barbeau <notifications@github.com> wrote:




@barbeau requested changes on this pull request.

Thanks @AnkitRajSri. Another comment in-line.

Also, the .pyc files are still under version control - you'll need to delete them and commit that.

@pradeepsalunke Were you able to clone this repository and run it locally on your machine without changing any code?

In config.py:
> @@ -0,0 +1,30 @@
+class Config:
+    """
+    Use this class to share any default attributes with any subsequent
+    classes that inherit from Config.
+    """
+    DEBUG = False
+    TESTING = False
+    
+    # Only required when using the session object
+    # Generated with secrets.token_urlsafe(16)
+    # You could also use os.urandom(16)
+    SECRET_KEY = "muser_data_app"

Did something change? - I still see the credentials above. They should be removed from the committed code and if necessary read it programmatically from an unversioned file. Need information in the README for how to configure this too.

**Comment by AnkitRajSri** [comment]
@barbeau I have deleted the .pyc files from the remote branch, hide the secret key, and added instructions for the config.py file in the README file.

@pradeepsalunke Let me know if you face any issues while testing the application on your local system.

**Comment by pradeepsalunke** [comment]
Hi Team,
I have tested the application locally and its working fine with few errors.

**Comment by pradeepsalunke** [comment]
1. Previous message doesn't get hidden even after clicking another button 
2. @AnkitRajSri is taking muser data in the form of a csv that we provided to expedite, it will be better if we can access the data directly from the muser app database, build it with the spotify features and then save as a SQL table.

**Comment by pradeepsalunke** [comment]
Hi Team,
I have tested application on my local application works as accepted except 3rd funcationlity which is to build the data.Below are the screen shotsPFA
Regards,Pradeep
Pradeep SalunkeGraduate Student,Masters in Business Analytics and Information SystemUniversity of South Florida


813.327.0938pradeep5@usf.edu


 

    On Monday, 21 December 2020, 08:35:01 PM IST, Ankit Raj <notifications@github.com> wrote:  
 
 


@barbeau I have deleted the .pyc files from the remote branch, hide the secret key, and added instructions for the config.py file in the README file.

@pradeepsalunke Let me know if you face any issues while testing the application on your local system.

**Comment by AnkitRajSri** [comment]
@barbeau @pradeepsalunke Sorry for the delayed response, I was traveling back to the US and couldn't look into your comment.
@pradeepsalunke thanks for the update, I am looking into the third functionality and will fix it asap.

**Comment by AnkitRajSri** [comment]
@barbeau @pradeepsalunke Seems like we were exhausting the Spotify API limit when we were using the muser data builder functionality after harvesting data from Spotify.
I wasn't able to replicate the issue when I tried to run just the data builder functionality.
I, however, have added a sleep condition to handle the API exhaustion, this should take care of the issue, even if we use all the functionalities together.

**Review by barbeau — APPROVED** [review]

**Comment by barbeau** [comment]
My understanding is that all the issues are resolved and @AnkitRajSri and @pradeepsalunke are both able to clone this PR and run from scratch on their own machines, with the exception of https://github.com/CUTR-at-USF/muser-data-analysis/issues/4, which doesn't always happen because it's related to an API quota. So, I'm going to go ahead and merge so we can iterate on the work in this PR.

**Pull request merged**